### PR TITLE
Add modular exponentiation(3-argument pow) for Expr

### LINKED ISSUE #5715: Ternary pow()
```
According to http://docs.python.org/release/2.5.2/ref/numeric-types.html , __pow__ should define a third, optional argument to work with ternary pow().  We should do that for at least Integer, though it would be cool to do it for arbitrary expressions (this will require Mod from issue 5589 , and also some care to make sure that it still evaluates efficiently when values are substituted in).  Right now, we get:

In [1]: pow(S(2), S(3), S(5))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)

/Users/aaronmeurer/Documents/python/sympy/sympy/<ipython console> in <module>()

TypeError: __sympifyit_wrapper() takes exactly 2 arguments (3 given)
```

Original issue for #5715: http://code.google.com/p/sympy/issues/detail?id=2616
Original author: https://code.google.com/u/asmeurer@gmail.com/
Referenced issues: #5589
Original owner: https://code.google.com/u/asmeurer@gmail.com/

### PR DESCRIPTION
**Pull request by ylemkimon**
Added modular exponentiation(3-argument pow) support for Expr.

Fixes #5715.

### CODE REVIEW

**Comment by Upabjojr** [comment]
I don't think this is the right way to implement this feature.

**Comment by Upabjojr** [comment]
This code should be inside Mod. You would probably want to call `Mod(a**b, m)`

**Comment by ylemkimon** [comment]
@Upabjojr Mod does that for unevaluated `Pow`. It was added in #13236.

**Comment by Upabjojr** [comment]
So maybe we need a new class, say `PowMod` or `ModPow`. Or a new method in `Pow`. I wouldn't hack `__pow__` to achieve this.

**Comment by ylemkimon** [comment]
@Upabjojr Python has built-in [3-argument pow(modular exponentiation)](https://docs.python.org/3/library/functions.html#pow).

**Comment by Upabjojr** [comment]
Oh, I see.

File: sympy/core/expr.py [hunk 1]
```diff
@@ -144,9 +144,25 @@ def __mul__(self, other):
...
+        except ValueError:
+            power = self.___pow__(other)
+            if power is NotImplemented:
+                return NotImplemented
+            return power%mod
+
     @_sympifyit('other', NotImplemented)
     @call_highest_priority('__rpow__')
-    def __pow__(self, other):
+    def ___pow__(self, other):
```
Comment (Upabjojr) [inline sympy/core/expr.py]: Why did you rename this?

Comment (ylemkimon) [reply sympy/core/expr.py]: @Upabjojr New `__pow__` calls `___pow__`, the original `__pow__`, if mod is None or not all arguments are integers. (See line 149 and 158 above)

Comment (Upabjojr) [reply sympy/core/expr.py]: I've seen this. But why not just keep one function?

Comment (ylemkimon) [reply sympy/core/expr.py]: @Upabjojr It’s because _sympifyit wrapper and call_highest_priority doesn’t support 3-argument function.

Comment (jksuom) [reply sympy/core/expr.py]: Perhaps this could be something like `_pow` as it is a private name. (Three underscores feels excessive:)

Comment (ylemkimon) [reply sympy/core/expr.py]: @jksuom Renamed ___pow__ to _pow and reordered functions.

File: sympy/core/expr.py [hunk 2]
```diff
@@ -146,9 +146,25 @@ def __rmul__(self, other):
...
+        try:
+            _self, other, mod = as_int(self), as_int(other), as_int(mod)
+            if other >= 0:
+                return pow(_self, other, mod)
+            else:
+                from sympy.core.numbers import mod_inverse
+                return mod_inverse(pow(_self, -other, mod), mod)
+        except ValueError:
+            power = self._pow(other)
+            if power is NotImplemented:
```
Comment (jksuom) [inline sympy/core/expr.py]: Can this really be true?

Comment (ylemkimon) [reply sympy/core/expr.py]: @jksuom `_pow` returns `NotImplemented` when _sympifying `other` fails, e.g., `pow(S(2), "2", 3)`. If we don't catch it, as `%`(mod) operation is not defined for `NotImplementedType`, it raises `TypeError: unsupported operand type(s) for %: 'NotImplementedType' and 'type(mod)'`, which I think is confusing and less informative. If we catch it and return `NotImplemented`, python raises `TypeError: unsupported operand type(s) for pow(): 'Integer', 'str', 'int'`.

Comment (smichr) [reply sympy/core/expr.py]: > it raises TypeError: unsupported operand type(s)

Then should the code be
```python
try:
    power = self._pow(other)
except TypeError:
    raise NotImplementedError  <--note NotImplementedError instead of NotImplemented?
```

And a test like `raises(NotImplementedError, lambda: pow(...failing args...))` should be added.

Comment (ylemkimon) [reply sympy/core/expr.py]: @smichr `_pow` itself does not raises TypeError, rather it returns `NotImplemented` when _sympify fails. Python automatically constructs and raises `TypeError` if `NotImplemented` is returned by overridden default operations, e.g., `__pow__`.

Comment (jksuom) [reply sympy/core/expr.py]: > _pow returns NotImplemented when _sympifying other fails, 

I didn't know that. I thought that `SympifyError` or `NotImplementedError` was raised if `other` is not sympifiable in `Pow(self, other)`.

Comment (smichr) [reply sympy/core/expr.py]: Can we get a test for this added to the suite so this line is covered?

File: sympy/core/tests/test_arit.py [hunk 3]
```diff
@@ -187,6 +187,13 @@ def test_pow3():
     assert sqrt(2)**3 == sqrt(8)
 
 
+def test_mod_pow():
+    assert pow(S(4), 13, 497) == pow(4, 13, 497)
+    assert pow(S(4), -3, 497)*4**3%497 == 1
+    assert pow(x, y, z) == x**y%z
```
Comment (asmeurer) [inline sympy/core/tests/test_arit.py]: Can you test all combinations of SymPy/Python input types? I don't remember exactly how Python does dispatch for three-argument `__pow__`, but we should make sure all the combinations are tested.

Comment (asmeurer) [reply sympy/core/tests/test_arit.py]: You might also add a test for something like `pow(S(2), S(10000000000), S(3))` (you can add more zeros if necessary), to make sure that it is using the efficient algorithm.

Comment (ylemkimon) [reply sympy/core/tests/test_arit.py]: @asmeurer I'm not sure how and whether to test that it is using efficient algorithm, since it depends on Python implementation.

Comment (asmeurer) [reply sympy/core/tests/test_arit.py]: `pow(S(2), S(10000000000), S(3))` will do the trick. If you try to evaluate `2**10000000000 % 3` naively it won't work. Using the Python implementation is fine. This just tests that it does indeed do it (or something equivalent).

File: sympy/core/expr.py (hunk 2 above)
Comment (ylemkimon) [reply sympy/core/expr.py]: @smichr Added 😄

Comment (ylemkimon) [reply sympy/core/expr.py]: Changed to catch TypeError, for when `mod` is not _sympifiable.

File: sympy/core/tests/test_arit.py (hunk 3 above)
Comment (ylemkimon) [reply sympy/core/tests/test_arit.py]: @asmeurer Added tests for various types 😃

**Comment by ylemkimon** [comment]
@asmeurer I missed your comment, sorry. Added test for very large numbers. But may there be a Python implementation other than Cython that does not implement efficient modular exponentiation?

**Comment by smichr** [comment]
@ylemkimon , does your question to Aaron mean that you want to hold off on committing this?

**Comment by ylemkimon** [comment]
@smichr No, it's good to merge. Thank you.

**Comment by asmeurer** [comment]
> But may there be a Python implementation other than CPython that does not implement efficient modular exponentiation?

I kind of doubt that. The whole point of having a three-argument pow() is to be able to compute modular exponents fast. The naive method could take forever to compute, or even crash the interpreter by running out of memory if `a**b` is too large. If a Python implementation implements three-argument pow() inefficiently a lot more code than just this test will fail under it.

**Comment by asmeurer** [comment]
Thanks. It's awesome to see this finally implemented.
